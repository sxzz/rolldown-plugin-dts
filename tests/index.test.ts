import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { rolldownBuild } from '@sxzz/test-utils'
import { describe, expect, test } from 'vitest'
import { dts } from '../src/index.ts'

const dirname = path.dirname(fileURLToPath(import.meta.url))

test('basic', async () => {
  const { snapshot } = await rolldownBuild(
    path.resolve(dirname, 'fixtures/basic.ts'),
    [dts()],
  )
  expect(snapshot).toMatchSnapshot()
})

test('tsx', async () => {
  const { snapshot } = await rolldownBuild(
    path.resolve(dirname, 'fixtures/tsx.tsx'),
    [dts()],
  )
  expect(snapshot).toMatchSnapshot()
})

test('resolve dependencies', async () => {
  const { snapshot } = await rolldownBuild(
    path.resolve(dirname, 'fixtures/resolve-dep.ts'),
    [
      dts({
        resolve: ['get-tsconfig'],
        oxc: true,
        emitDtsOnly: true,
      }),
    ],
  )
  expect(snapshot).contain('type TsConfigResult')
  expect(snapshot).not.contain('node_modules/rolldown')
})

test('resolve dts', async () => {
  const { snapshot } = await rolldownBuild(
    path.resolve(dirname, 'fixtures/resolve-dts/index.ts'),
    [dts()],
  )
  expect(snapshot).matchSnapshot()
})

// Test alias mapping based on rolldown input option
test('input alias', async () => {
  const root = path.resolve(dirname, 'fixtures/alias')
  const { snapshot, chunks } = await rolldownBuild(
    null!,
    [dts({ emitDtsOnly: false })],
    {
      cwd: root,
      // A mapping from output chunk names to input files. This mapping should
      // be used in both JS and DTS outputs.
      input: {
        output1: 'input1.ts',
        'output2/index': 'input2.ts',
      },
    },
  )
  const fileNames = chunks.map((chunk) => chunk.fileName).sort()

  // The JS output and DTS output should have the same structure
  expect(fileNames).toContain('output1.d.ts')
  expect(fileNames).toContain('output1.js')
  expect(fileNames).toContain('output2/index.d.ts')
  expect(fileNames).toContain('output2/index.js')

  expect(snapshot).toMatchSnapshot()
})

test('isolated declaration error', async () => {
  const error = await rolldownBuild(
    path.resolve(dirname, 'fixtures/isolated-decl-error.ts'),
    [
      dts({
        emitDtsOnly: true,
        oxc: true,
      }),
    ],
  ).catch((error: any) => error)
  expect(String(error)).toContain(
    `Function must have an explicit return type annotation with --isolatedDeclarations.`,
  )
  expect(String(error)).toContain(`export function fn() {`)
})

test('paths', async () => {
  const root = path.resolve(dirname, 'fixtures/paths')
  const { snapshot } = await rolldownBuild(path.resolve(root, 'index.ts'), [
    dts({
      oxc: true,
      emitDtsOnly: true,
      tsconfig: path.resolve(root, 'tsconfig.json'),
    }),
  ])
  expect(snapshot).toMatchSnapshot()
})

test('tree-shaking', async () => {
  const { snapshot } = await rolldownBuild(
    path.resolve(dirname, 'fixtures/tree-shaking/index.ts'),
    [
      dts(),
      {
        name: 'external-node',
        resolveId(id) {
          if (id.startsWith('node:'))
            return { id, external: true, moduleSideEffects: false }
        },
      },
    ],
    { treeshake: true },
  )
  expect(snapshot).matchSnapshot()
})

describe('dts input', () => {
  test('input array', async () => {
    const { snapshot, chunks } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/dts-input.d.ts')],
      [dts({ dtsInput: true })],
      {},
    )
    expect(chunks[0].fileName).toBe('dts-input.d.ts')
    expect(snapshot).toMatchSnapshot()
  })

  test('input object', async () => {
    const { snapshot, chunks } = await rolldownBuild(
      null!,
      [dts({ dtsInput: true })],
      {
        input: {
          index: path.resolve(dirname, 'fixtures/dts-input.d.ts'),
        },
      },
    )
    expect(chunks[0].fileName).toBe('index.d.ts')
    expect(snapshot).toMatchSnapshot()
  })

  test('.d in chunk name', async () => {
    const { chunks } = await rolldownBuild(null!, [dts({ dtsInput: true })], {
      input: {
        'index.d': path.resolve(dirname, 'fixtures/dts-input.d.ts'),
      },
    })
    expect(chunks[0].fileName).toBe('index.d.ts')
  })

  test('full extension in chunk name', async () => {
    const { chunks } = await rolldownBuild(null!, [dts({ dtsInput: true })], {
      input: {
        'index.d.mts': path.resolve(dirname, 'fixtures/dts-input.d.ts'),
      },
    })
    expect(chunks[0].fileName).toBe('index.d.mts')
  })

  test('custom entryFileNames with .d', async () => {
    const { chunks } = await rolldownBuild(
      null!,
      [dts({ dtsInput: true })],
      {
        input: {
          index: path.resolve(dirname, 'fixtures/dts-input.d.ts'),
        },
      },
      {
        entryFileNames: '[name].d.cts',
      },
    )
    expect(chunks[0].fileName).toBe('index.d.cts')
  })

  test('custom entryFileNames without .d', async () => {
    const { chunks } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/dts-input.d.ts')],
      [dts({ dtsInput: true })],
      {},
      {
        entryFileNames: '[name].mts',
      },
    )
    expect(chunks[0].fileName).toBe('dts-input.d.mts')
  })

  test('custom entryFileNames function', async () => {
    const { chunks } = await rolldownBuild(
      null!,
      [dts({ dtsInput: true })],
      {
        input: {
          index: path.resolve(dirname, 'fixtures/dts-input.d.ts'),
        },
      },
      {
        entryFileNames: () => '[name].mts',
      },
    )
    expect(chunks[0].fileName).toBe('index.d.mts')
  })

  test('invalid entryFileNames gets overridden with stripped .d', async () => {
    const { chunks } = await rolldownBuild(
      null!,
      [dts({ dtsInput: true })],
      {
        input: {
          'index.d': path.resolve(dirname, 'fixtures/dts-input.d.ts'),
        },
      },
      {
        entryFileNames: '[name].invalid',
      },
    )
    expect(chunks[0].fileName).toBe('index.d.ts')
  })

  test('invalid entryFileNames gets overridden and preserves subextension', async () => {
    const { chunks } = await rolldownBuild(
      null!,
      [dts({ dtsInput: true })],
      {
        input: {
          'index.asdf': path.resolve(dirname, 'fixtures/dts-input.d.ts'),
        },
      },
      {
        entryFileNames: '[name].invalid',
      },
    )
    expect(chunks[0].fileName).toBe('index.asdf.d.ts')
  })

  test('default chunk name', async () => {
    const { snapshot, chunks } = await rolldownBuild(
      [
        path.resolve(dirname, 'fixtures/dts-multi-input/input1.d.ts'),
        path.resolve(dirname, 'fixtures/dts-multi-input/input2.d.ts'),
      ],
      [dts({ dtsInput: true })],
      {},
      {
        entryFileNames: '[name].mts',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual([
      'input1.d.mts',
      'input2.d.mts',
      'types-VwSK8P_f.d.ts',
    ])

    expect(snapshot).toMatchSnapshot()
  })

  test('custom chunk name', async () => {
    const { snapshot, chunks } = await rolldownBuild(
      [
        path.resolve(dirname, 'fixtures/dts-multi-input/input1.d.ts'),
        path.resolve(dirname, 'fixtures/dts-multi-input/input2.d.ts'),
      ],
      [dts({ dtsInput: true })],
      {},
      {
        chunkFileNames: 'chunks/[hash]-[name].ts',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual([
      'chunks/DqALGAwS-types.d.ts',
      'input1.d.ts',
      'input2.d.ts',
    ])

    expect(snapshot).toMatchSnapshot()
  })
})

describe('entryFileNames', () => {
  test('.mjs -> .d.mts', async () => {
    const { chunks } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/basic.ts')],
      [dts()],
      {},
      {
        entryFileNames: '[name].mjs',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual(['basic.d.mts', 'basic.mjs'])
  })

  test('.cjs -> .d.cts', async () => {
    const { chunks } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/basic.ts')],
      [dts()],
      {},
      {
        entryFileNames: '[name].cjs',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual(['basic.cjs', 'basic.d.cts'])
  })

  test('.mjs -> .d.mts with custom chunk name', async () => {
    const { chunks } = await rolldownBuild(
      null!,
      [dts()],
      {
        input: {
          custom: path.resolve(dirname, 'fixtures/basic.ts'),
        },
      },
      {
        entryFileNames: '[name].mjs',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual(['custom.d.mts', 'custom.mjs'])
  })

  test('preserves invalid extension', async () => {
    const { chunks } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/basic.ts')],
      [dts()],
      {},
      {
        entryFileNames: '[name].invalid',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual(['basic.d.invalid', 'basic.invalid'])
  })

  test('same-name output (for JS & DTS)', async () => {
    const { chunks } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/same-name/index.ts')],
      [dts()],
      {},
      {
        preserveModules: true,
        entryFileNames: 'foo.d.ts',
      },
    )

    expect(chunks.every((chunk) => chunk.fileName.endsWith('.d.ts'))).toBe(true)
  })

  test('default chunk name', async () => {
    const { chunks, snapshot } = await rolldownBuild(
      [
        path.resolve(dirname, 'fixtures/alias/input1.ts'),
        path.resolve(dirname, 'fixtures/alias/input2.ts'),
      ],
      [dts({ emitDtsOnly: true })],
      {},
      {
        entryFileNames: '[name].mjs',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual([
      'input1.d.mts',
      'input2-CzdQ8V-e.d.ts',
      'input2.d.mts',
    ])

    expect(snapshot).toMatchSnapshot()
  })

  test('custom chunk name', async () => {
    const { snapshot, chunks } = await rolldownBuild(
      [
        path.resolve(dirname, 'fixtures/dts-multi-input/input1.d.ts'),
        path.resolve(dirname, 'fixtures/dts-multi-input/input2.d.ts'),
      ],
      [dts({ emitDtsOnly: true })],
      {},
      {
        chunkFileNames: 'chunks/[hash]-[name].js',
      },
    )

    const chunkNames = chunks.map((chunk) => chunk.fileName).sort()
    expect(chunkNames).toStrictEqual([
      'chunks/DqALGAwS-types.d.ts',
      'input1.d.ts',
      'input2.d.ts',
    ])

    expect(snapshot).toMatchSnapshot()
  })
})

test('type-only export', async () => {
  const { snapshot } = await rolldownBuild(
    [path.resolve(dirname, 'fixtures/type-only-export/index.ts')],
    [dts({ emitDtsOnly: true })],
  )
  expect(snapshot).toMatchSnapshot()
})

test('cjs exports', async () => {
  {
    const { snapshot } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/cjs-exports.ts')],
      [],
      {},
      { format: 'cjs', exports: 'auto' },
    )
    expect(snapshot).toMatchSnapshot()
  }

  {
    const { snapshot } = await rolldownBuild(
      [path.resolve(dirname, 'fixtures/cjs-exports.ts')],
      [dts({ emitDtsOnly: true, cjsDefault: true })],
    )
    expect(snapshot).toMatchSnapshot()
  }
})

test('should error when file import cannot be found', async () => {
  await expect(() =>
    rolldownBuild(path.resolve(dirname, 'fixtures/unresolved-import/ts.ts'), [
      dts({
        emitDtsOnly: true,
      }),
    ]),
  ).rejects.toThrow("Could not resolve './missing-file'")
})
