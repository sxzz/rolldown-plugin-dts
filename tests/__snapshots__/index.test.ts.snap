// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
"// basic.d.ts
//#region tests/fixtures/basic.d.ts
declare const foo: number;
type SomeType<T> = T;
type FooType = string;
interface Interface {}
declare function fn(arg0: SomeType<FooType>, opt: Interface): void;
declare enum Enum {
  A = 0,
  B = 1,
  C = 2,
}
declare class Cls {
  foo: string;
  fn(e: Enum): void;
}
//#endregion
export { Cls, Enum, fn, foo };
// basic.js
//#region tests/fixtures/basic.ts
const foo = 42;
function fn(arg0, opt) {}
let Enum = /* @__PURE__ */ function(Enum$1) {
	Enum$1[Enum$1["A"] = 0] = "A";
	Enum$1[Enum$1["B"] = 1] = "B";
	Enum$1[Enum$1["C"] = 2] = "C";
	return Enum$1;
}({});
var Cls = class {
	foo;
	fn(e) {}
};

//#endregion
export { Cls, Enum, fn, foo };"
`;

exports[`dts input 1`] = `
"// index.d.ts
//#region tests/fixtures/dts-input.d.ts
declare class Foo {}

//#endregion
export { Foo };"
`;

exports[`input alias 1`] = `
"// input2-DX9os_AR.d.ts
//#region shared.d.ts
interface Shared {
  shared: string;
}
declare const shared: Shared;
//#endregion
//#region input2.d.ts
interface Input2 extends Shared {
  input2: string;
}
declare const input2: Input2;

//#endregion
export { Input2, input2 as input2$1 };
// input2-LCiYuoaZ.js
//#region shared.ts
const shared = { shared: "shared" };

//#endregion
//#region input2.ts
const input2 = {
	...shared,
	input2: "input2"
};

//#endregion
export { input2 };
// output1.d.ts
import { Input2 } from "./input2-DX9os_AR.js";

//#region input1.d.ts
interface Input1 extends Input2 {
  input1: string;
}
declare const input1: Input1;

//#endregion
export { Input1, input1 };
// output1.js
import { input2 } from "./input2-LCiYuoaZ.js";

//#region input1.ts
const input1 = {
	...input2,
	input1: "input1"
};

//#endregion
export { input1 };
// output2/index.d.ts
import { Input2, input2$1 as input2 } from "../input2-DX9os_AR.js";
export { Input2, input2 };
// output2/index.js
import { input2 } from "../input2-LCiYuoaZ.js";

export { input2 };"
`;

exports[`jsdoc 1`] = `
"// jsdoc.d.ts
//#region tests/fixtures/jsdoc.d.ts
/**
 * named export
 */
declare function fn(): number;
/**
 * options
 */
interface Options {
  /**
   * interface member
   */
  foo: string;
}
/**
 * type alias
 */
type Foo = string;
/**
 * default export
 */
declare const _default: number; //#endregion
export { Foo, Options, _default as default, fn };
// jsdoc.js
//#region tests/fixtures/jsdoc.ts
/**
* named export
*/
function fn() {
	return 42;
}
/**
* default export
*/
var jsdoc_default = fn();

//#endregion
export { jsdoc_default as default, fn };"
`;

exports[`paths 1`] = `
"// index.d.ts
//#region tests/fixtures/paths/mod.d.ts
declare const foo = 42;

//#endregion
export { foo };"
`;

exports[`resolve dts 1`] = `
"// index.d.ts
//#region tests/fixtures/resolve-dts/mod.d.ts
type Foo = string;

//#endregion
export { Foo };
// index.js
"
`;

exports[`tree-shaking 1`] = `
"// index.d.ts
//#region tests/fixtures/tree-shaking/mod.d.ts

type A = string;
//#endregion
export { A };
// index.js
"
`;

exports[`tsx 1`] = `
"// tsx.d.ts
import { JSX } from "solid-js";

//#region tests/fixtures/tsx.d.ts
declare function createComponent(): JSX.HTMLAttributes<HTMLElement>;

//#endregion
export { createComponent };
// tsx.js
import { jsx } from "react/jsx-runtime";

//#region tests/fixtures/tsx.tsx
function createComponent() {
	return /* @__PURE__ */ jsx("div", {});
}

//#endregion
export { createComponent };"
`;

exports[`typescript compiler 1`] = `
"// entry1.d.ts

declare const str = "foo";
declare const num = 42;
declare const bool = true;
declare function fn(arg: typeof str): "foo" | 1;

//#endregion

declare const vfs: number;

//#endregion
export { bool, fn, num, str, vfs };
// entry2.d.ts

interface Unused {
  foo: string;
}
//#endregion
export { Unused };"
`;

exports[`vue-sfc w/ ts-compiler 1`] = `
"// main.d.ts
//#region tests/fixtures/vue-sfc/App.vue.d.ts
declare global {
  interface Window {
    foo: string;
  }
}
declare const _default: any; //#endregion
export { _default as App };"
`;
